osu.Game/Screens/Edit/Timing/WaveformComparisonDisplay.cs

```diff
@@ -156,23 +199,83 @@ private void showFrom(int beatIndex)
             float trackLength = (float)beatmap.Value.Track.Length;
             float scale = trackLength / visible_width;
 
+            const int start_offset = total_waveforms / 2;
+
             // Start displaying from before the current beat
-            beatIndex -= total_waveforms / 2;
+            index -= start_offset;
 
             foreach (var row in InternalChildren.OfType<WaveformRow>())
             {
                 // offset to the required beat index.
-                double time = selectedGroupStartTime + beatIndex * timingPoint.BeatLength;
+                double time = selectedGroupStartTime + index * timingPoint.BeatLength;
 
                 float offset = (float)(time - visible_width / 2) / trackLength * scale;
 
                 row.Alpha = time < selectedGroupStartTime || time > selectedGroupEndTime ? 0.2f : 1;
-                row.WaveformOffset = -offset;
+                row.WaveformOffsetTo(-offset, animated);
                 row.WaveformScale = new Vector2(scale, 1);
-                row.BeatIndex = beatIndex++;
+                row.BeatIndex = (int)Math.Floor(index);
```

Comment:
It confused me for a good while why `index` is permitted to be fractional all the way up to this point (thinking that the display was kind of "lying" by pretending that it's showing waveforms for beats, while it is actually showing parts of neighbouring bars glued together). After trying to floor the index more aggressively I get why that is the case now (leads to smoother display when locked, and the waveform will resnap to a correct beat on next `Update()` after unlocking), but maybe worth adding an inline comment?

Reply:
It has to be fractional to support display at arbitrary time. Else that pathway would not work correctly.

I considered rewriting this method to not use beat index but it didn't seem worthwhile.

Reply:
I get why it has to be fractional, it's just more that it felt like a bit of a lie to be showing integral beat indices while the current time instant is not snapped to a beat. I'd honestly consider hiding the indices in the mean time maybe?